### PR TITLE
[perf] reduce D2H copies and allocations in DP attention scheduler sync

### PR DESCRIPTION
**Pull request by alexnails**
## Summary

Optimizes the per-step CPU overhead in `prepare_mlp_sync_batch_raw` / `MLPSyncBatchInfo.all_gather` on the DP attention scheduler path. Four changes, all in `scheduler_dp_attn_mixin.py` and `two_batch_overlap.py`:

- **`int32` instead of `int64`** for the all-gather info tensor — halves the NCCL wire size from 384 B → 192 B per all-gather (6 fields × 8 ranks × dtype)
- **Pre-allocated `_MLPSyncBuffers` singleton** — `local_tensor`, `global_tensor`, `global_tensor_flat`, and `fallback_tensor` are allocated once at first call and reused in-place each step, eliminating per-step CUDA/CPU malloc
- **Cached `TboDPAttentionPreparer`** — module-level singleton instead of constructing a new object every scheduler step
- **4 D2H copies → 1** — previously `tp0_info[:, :2].cpu()`, `.min().item()`, `.max().item()`, and `compute_output`'s `.cpu()` were separate device-to-host transfers; now a single `tp0_info.cpu()` copies all 6 columns at once, and `cpu_tp0_info` is threaded through so `compute_output` needs no extra copy
- **Fast path for inactive ranks** — skips the fallback-value write when all ranks are active (the common homogeneous case)

## Benchmark

**Setup:** Qwen3-30B-A3B-Instruct-2507, tp=4, dp=2, ep=4, `--enable-dp-attention`, 8×H200  
**Workload:** 500 prompts, 512 input / 512 output tokens, max-concurrency=200

| Metric | `main` | This PR | Δ |
|--------|--------|---------|---|
| Output token throughput | 5762 tok/s | **5814 tok/s** | **+0.9%** |
| Total token throughput | 11330 tok/s | **11431 tok/s** | **+0.9%** |
| Request throughput | 21.88 req/s | **22.08 req/s** | **+0.9%** |
| Benchmark duration | 22.85 s | **22.65 s** | **-0.9%** |
| Mean E2E latency | 7923 ms | **7845 ms** | **-1.0%** |
| Mean TTFT | 289 ms | **256 ms** | **-11.4%** |
| P99 TTFT | 697 ms | **628 ms** | **-9.9%** |
| Mean TPOT | 30.72 ms | **30.50 ms** | **-0.7%** |
| P99 TPOT | 47.39 ms | **44.86 ms** | **-5.3%** |
| Max ITL | 539 ms | **478 ms** | **-11.2%** |

TTFT and tail latency see the largest gains (~10%) because the scheduler decides faster each step once the 4→1 D2H bottleneck is removed, reducing time-to-first-token and worst-case per-token pipeline bubbles.

**Profiler evidence (TP-0 rank, same workload):**
- `ncclDevKernel_AllGather` per-launch: 10.7 µs → 7.45 µs (−28%), directly attributable to the int32 wire-size reduction
- `cross_device_reduce_1stage` accumulated: 135 ms → 3.7 ms, reflecting the GPU pipeline no longer stalling behind back-to-back D2H transfers in the scheduler

## Correctness

- All 15 requests passed (5 sequential + 10 concurrent) on Qwen3-30B-A3B with dp-attn enabled
- `int32` overflow analysis: largest field is `num_tokens`; even at 128k context × 1000 concurrent requests = 128M, well within int32 max (~2.1B)
- `_MLPSyncBuffers` singleton is safe: `_fill_local_tensor` overwrites all 6 fields in-place each step before the all-gather reads them; no stale state possible
- `_tbo_preparer` singleton is safe: `prepare_all_gather` sets `self.enable_two_batch_overlap` and `self.local_tbo_split_seq_index` fresh each step before `compute_output` reads them

### CODE REVIEW

File: python/sglang/srt/managers/scheduler_dp_attn_mixin.py [hunk 1]
```diff
@@ -20,6 +20,82 @@
 
 _ENABLE_METRICS_DP_ATTENTION = envs.SGLANG_ENABLE_METRICS_DP_ATTENTION.get()
 
+# int32 suffices for all 6 packed fields (token counts, bools, forward-mode enum)
+# and halves the all_gather wire size vs int64.
+_INFO_DTYPE = torch.int32
```
Comment (ch-wan) [inline python/sglang/srt/managers/scheduler_dp_attn_mixin.py:25]:
```suggestion
_MLP_SYNC_INFO_DTYPE = torch.int32
```

File: python/sglang/srt/batch_overlap/two_batch_overlap.py [hunk 2]
```diff
@@ -417,10 +417,10 @@ def prepare_all_gather(
         return local_can_run_tbo, local_forward_mode
 
     def compute_output(self, partial_global_info):
-        # Perform only one Device-to-Host (D2H) memory copy
-        cpu_data = partial_global_info[:, :2].cpu()
-        local_can_run_tbo_aggregated = min(cpu_data[:, 0].tolist())
-        forward_modes = cpu_data[:, 1].tolist()
+        # partial_global_info is already on CPU (caller passes cpu_tp0_info[:, 4:6]);
+        # no D2H copy is needed here.
```
Comment (ch-wan) [inline python/sglang/srt/batch_overlap/two_batch_overlap.py:421]: comment here is not needed

File: python/sglang/srt/managers/scheduler_dp_attn_mixin.py [hunk 3]
```diff
@@ -20,6 +20,82 @@
 
 _ENABLE_METRICS_DP_ATTENTION = envs.SGLANG_ENABLE_METRICS_DP_ATTENTION.get()
 
+# int32 suffices for all 6 packed fields (token counts, bools, forward-mode enum)
+# and halves the all_gather wire size vs int64.
```
Comment (ch-wan) [inline python/sglang/srt/managers/scheduler_dp_attn_mixin.py:24]: This comment is part of the pr description, not code.

File: python/sglang/srt/managers/scheduler_dp_attn_mixin.py [hunk 4]
```diff
@@ -20,6 +20,82 @@
...
+        _tbo_preparer = TboDPAttentionPreparer()
+    return _tbo_preparer
+
+
+def _get_or_init_mlp_sync_buffers(
+    dp_size: int,
+    tp_size: int,
+    cp_size: int,
+    device,
+    dtype: torch.dtype,
```
Comment (ch-wan) [inline python/sglang/srt/managers/scheduler_dp_attn_mixin.py:62]: Passing `dtype` is not redundant. We can use info dtype directly.

File: python/sglang/srt/managers/scheduler_dp_attn_mixin.py [hunk 5]
```diff
@@ -20,6 +20,82 @@
...
+            pin_memory=use_pin,
+        )
+        _mlp_sync_buffers = _MLPSyncBuffers(
+            local_tensor=torch.zeros(6, dtype=dtype, device=device),
+            global_tensor=global_tensor,
+            global_tensor_flat=global_tensor.flatten(),
+            fallback_tensor=torch.tensor(fallback_values, dtype=dtype, device=device),
+            has_inactive_ranks=has_inactive,
+        )
+    return _mlp_sync_buffers
```
Comment (ch-wan) [inline python/sglang/srt/managers/scheduler_dp_attn_mixin.py:97]: Can we add some assertions to check if this buffer's size and device match the input?

File: python/sglang/srt/managers/scheduler_dp_attn_mixin.py [hunk 6]
```diff
@@ -36,72 +112,57 @@ class MLPSyncBatchInfo:
 
     # some gathered elements
     tp0_info: torch.Tensor = None
+    cpu_tp0_info: torch.Tensor = None
```
Comment (ch-wan) [inline python/sglang/srt/managers/scheduler_dp_attn_mixin.py:115]: `tp0_info_cpu`
